src/BSeries.jl

```diff
@@ -1453,4 +1456,419 @@ include("latexify.jl")
+end
+
+"""
+This functions checks if the level_sequence is a bush.
+"""
+function bush_detector(tree)
+    bush = false
+    if tree != [1]
+        l = length(tree)
+        bush = true
+        initial_index = tree[1]
+        initial_plus_one = initial_index + 1
+        if l > 1
+            for i in 2:l
+                if tree[i] != initial_plus_one
+                    bush = false
+                end
+            end
+        end
+    end
+    return bush
+end
+
+#this function eliminates repeated subarrays
+function eliminate_repeated_subarrays(M)
+    unique_M = []
+    for arr in M
+        if arr in unique_M
+        else
+            push!(unique_M, arr)
+        end
+    end
+    return unique_M
+end
+
+#this function swaps the trees
+function permuta(a::Vector{Int})
+    #generate all the leafs via 'modify_t_sub'
+    diccionario = modify_t_sub(a)
+    m = num_leafs(a)
+    ad_dict = Dict{Int, Vector{Int}}()
+    #we swap every leaf j for the (m-j+1)
+    for j in 1:m
+        ad_dict[j] = diccionario[m-j+1]
+    end
+    return ad_dict
+end
+
+"""
+This function returns the rightmost_energy_preserving_tree level_sequence 
+for a given tree (the input also in level-sequence form).
+"""
+function rightmost_energy_preserving_tree(a::Vector{Int})
+    #we want to generate all the leafs with respect to the rightmost spine
+    ad_dict = permuta(a)
+    #we obtain the number of leafs the right-most spine has
+    #the Theorem 2 in the article requires to know if m is odd or even
+    m = num_leafs(a)
+    #we create an array from 1 to m plus another node for the final leaf of the rightmost spine
+    adjunto = collect(1:m+1)
+    #then, we insert every level_sequence from ad_dict
+    for j in 1:m
+        last_j_occurrence = findlast(x -> x == j, adjunto)
+        last_jplus1_occurrence = findlast(x -> x == j+1, adjunto)
+        adjunto = vcat(adjunto[1:last_j_occurrence], ad_dict[j], adjunto[last_jplus1_occurrence:end])
+    end
+    adjunto = canonicalarray(adjunto)
+    #println("Adjoint Array: ", adjunto)
+    return adjunto
+end
+
+"""
+This function rearranges the level sequence in 
+the same way as the Bseries output does
+"""
+function canonicalarray(tree)
+    t = rootedtree(tree)
+    trees = t.level_sequence
+    return trees
+end
+
+"""
+        indexator(trees,onetree)
+This functions receives as inputs an array-of-arrays and 
+one of its elements.
+
+The output is the index of this array 'onetree'.
+"""
+function indexator(trees,onetree)
+    theindex = 0
+    l = length(trees)
+    for i in 1:l
+        if trees[i] == onetree
+            theindex = i
+        end
+    end
+    return theindex
+end
+
+iswhitespace(c::Char) = isspace(c)
+   
+
+"""
+This function generates all the equivalent trees 
+for a given level_sequence
+"""
+function equivalent_trees(array)
+    tree = rootedtree(array)
+    l = length(array)
+    #we get the permutations of the array
+    superarray = get_permutations(array)
```

Comment:
This is the main bottleneck:
```julia
julia> using ProfileView, BSeries

julia> series = bseries(8) do t, series
           if order(t) in (0, 1)
               return 1 // 1
           else
               v = 1 // 1
               n = 0
               for subtree in SubtreeIterator(t)
                   v *= series[subtree]
                   n += 1
               end
               return v / (n + 1)
           end
       end

julia> @profview is_energy_preserving(series) # result of calling this the second time, after compilation
```

![image](https://github.com/ranocha/BSeries.jl/assets/12693098/5da2ab3f-f42c-4465-8860-924143cfa9d0)

When you refine the implementation, you should consider concentrating on this. 
- Instead of saving all permutations first and throwing away most of them later, could you just iterate over all permutations lazily and just keep the ones that are required?
- Is there another algorithm with reduced complexity to generate all equivalent representation of a rooted tree?